### PR TITLE
Move sorters in their own package; Adding sorters

### PR DESCRIPTION
**Pull request by draychev**
We talked about sorting lists of config in relation to better caching.
I did not intend on working on this, until I realized it would be very helpful for debugging (while working on multiple namespace support etc) to compare config changes between PUTs.
It turns out the more complex the config, the higher the churn in config -- JSON changes quite a lot and it is difficult to compare.

This PR here introduces a bunch of sorters, which are applied to the respective lists of structs before the addresses of these structs are assigned to the Application Gateway config.

### CODE REVIEW

File: pkg/sorter/certificates.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// ByCertificateName is a facility to sort slices of ApplicationGatewaySslCertificate by Name
+type ByCertificateName []n.ApplicationGatewaySslCertificate
+
+func (a ByCertificateName) Len() int      { return len(a) }
+func (a ByCertificateName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByCertificateName) Less(i, j int) bool {
+	return getCertificateName(a[i]) < getCertificateName(a[j])
+}
+
+func getCertificateName(rule n.ApplicationGatewaySslCertificate) string {
```
Comment (akshaysngupta) [inline pkg/sorter/certificates.go]: cert

File: pkg/sorter/frontend_ports.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+type ByFrontendPortName []n.ApplicationGatewayFrontendPort
+
+func (a ByFrontendPortName) Len() int      { return len(a) }
+func (a ByFrontendPortName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByFrontendPortName) Less(i, j int) bool {
+	return getFrontendPortName(a[i]) < getFrontendPortName(a[j])
+}
+
+func getFrontendPortName(rule n.ApplicationGatewayFrontendPort) string {
+	if rule.Name == nil {
```
Comment (akshaysngupta) [inline pkg/sorter/frontend_ports.go]: `port`

File: pkg/sorter/listeners.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+type ByListenerName []n.ApplicationGatewayHTTPListener
+
+func (a ByListenerName) Len() int      { return len(a) }
+func (a ByListenerName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByListenerName) Less(i, j int) bool {
+	return getListenerName(a[i]) < getListenerName(a[j])
+}
+
+func getListenerName(rule n.ApplicationGatewayHTTPListener) string {
+	if rule.Name == nil {
```
Comment (akshaysngupta) [inline pkg/sorter/listeners.go]: s/`rule`/`listener`

File: pkg/sorter/path_maps.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+type ByPathMap []n.ApplicationGatewayURLPathMap
+
+func (a ByPathMap) Len() int      { return len(a) }
+func (a ByPathMap) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByPathMap) Less(i, j int) bool {
+	return getPathMapName(a[i]) < getPathMapName(a[j])
+}
+
+func getPathMapName(rule n.ApplicationGatewayURLPathMap) string {
+	if rule.Name == nil {
```
Comment (akshaysngupta) [inline pkg/sorter/path_maps.go]: s/`rule`/`pathMap`

File: pkg/sorter/redirects.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+type ByRedirectName []n.ApplicationGatewayRedirectConfiguration
+
+func (a ByRedirectName) Len() int      { return len(a) }
+func (a ByRedirectName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByRedirectName) Less(i, j int) bool {
+	return getRedirectName(a[i]) < getRedirectName(a[j])
+}
+
+func getRedirectName(rule n.ApplicationGatewayRedirectConfiguration) string {
+	if rule.Name == nil {
```
Comment (akshaysngupta) [inline pkg/sorter/redirects.go]: s/`rule`/`redirect`

**Comment by draychev** [comment]
For the record - next possible step in simplifying this would be to create a common interface; typealias SDK's structs; reuse the same getName function:


```go
import (
	n "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network"
)

type Config interface {
	GetName() *string
}

type RedirectConfiguration n.ApplicationGatewayRedirectConfiguration

func (c RedirectConfiguration) GetName() *string {
	return c.Name
}

// ----------

type Certificate n.ApplicationGatewaySslCertificate

func (c Certificate) GetName() *string {
	return c.Name
}

type FrontendPort n.ApplicationGatewayFrontendPort

func (c FrontendPort) GetName() *string {
	return c.Name
}

type Probe n.ApplicationGatewayProbe

func (c Probe) GetName() *string {
	return c.Name
}

type Listener n.ApplicationGatewayHTTPListener

func (c Listener) GetName() *string {
	return c.Name
}

type PathMap n.ApplicationGatewayURLPathMap

func (c PathMap) GetName() *string {
	return c.Name
}

// ----------

func getName(config Config) string {
	if config.GetName() == nil {
		return ""
	}
	return *config.GetName()
}
```

**Pull request merged**